### ISSUE TITLE
Filter gets mis-optimized and doesn't filter out false values

### ISSUE THREAD

**Issue by TheDan64** [pola-rs/polars#1395]
#### Are you using Python or Rust?

Rust

#### Which feature gates did you use?

`"lazy", "random", "temporal", "dtype-date64", "dtype-u64", "object"`

#### What version of polars are you using?

Latest or nearly latest master (`06448f61bea22f1a7407bbfa1c3a364354192713`)

#### What operating system are you using polars on?

Win10

#### Describe your bug.

This particular filter by a boolean column filters out null values but not false ones.

#### What are the steps to reproduce the behavior?

```rust
fn foo(lf: LazyFrame, column_name: &str) -> LazyFrame {
    let shift_col_1 = col(column_name).shift_and_fill(1, lit(true)).lt(col(column_name));
    let shift_col_neg_1 = col(column_name).shift(-1).lt(col(column_name));
    let filled_shift_1 = col("shift_1");
    let filled_shift_neg_1 = col("shift_neg_1");
    let filled_shift_1_2 = col("shift_1_2");
    let filled_shift_neg_1_2 = col("shift_neg_1_2").fill_null(lit(true));
    lf
        .with_columns(vec![
            shift_col_1.clone().alias("shift_1"),
            shift_col_neg_1.clone().alias("shift_neg_1"),
        ])
        .with_column(filled_shift_1.and(filled_shift_neg_1).alias("diff"))
        .filter(col("diff"))
}

pub fn test() -> Result<()> {
    let dts = vec![
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 0, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 1, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 2, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 3, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 4, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 5, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 6, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 7, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 8, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 9, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 10, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 11, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 12, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 13, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 14, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 15, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 16, 0),
        NaiveDate::from_ymd(2021, 11, 11).and_hms(11, 17, 0),
    ];
    let data = vec![
        None, None, None, None, Some(false), Some(false), Some(true), Some(false), Some(true),
        Some(false), Some(true), Some(false), Some(true), Some(false), Some(true), Some(false), Some(false), None
    ];
    let series = Series::new("data", data);
    let df = DataFrame::new(vec![
        Date64Chunked::new_from_naive_datetime("date", &*dts).into(),
        series,
    ])?;

    dbg!(foo(df.lazy(), "data").collect());

    Ok(())
}
```

#### What is the actual behavior?

This produces rows with false diff column values:

```rust
    shape: (13, 5)
    +---------------------+-------+---------+-------------+-------+
    | date                | data  | shift_1 | shift_neg_1 | diff  |
    | ---                 | ---   | ---     | ---         | ---   |
    | date64(ms)          | bool  | bool    | bool        | bool  |
    +=====================+=======+=========+=============+=======+
    | 2021-11-11 11:04:00 | false | null    | false       | false |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:05:00 | false | false   | false       | false |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:06:00 | true  | true    | true        | true  |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:07:00 | false | false   | false       | false |
    +---------------------+-------+---------+-------------+-------+
    | ...                 | ...   | ...     | ...         | ...   |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:12:00 | true  | true    | true        | true  |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:13:00 | false | false   | false       | false |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:14:00 | true  | true    | true        | true  |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:15:00 | false | false   | false       | false |
    +---------------------+-------+---------+-------------+-------+
    | 2021-11-11 11:16:00 | false | false   | null        | false |
    +---------------------+-------+---------+-------------+-------+
```

#### What is the expected behavior?

It is expected that `false` values would get filtered out

**Comment by ritchie46** [comment]
Thanks @TheDan64. I will take a look.

**Comment by ritchie46** [comment]
There was a false positive in an optimization rule. This is fixed in #1398

**Comment by TheDan64** [comment]
Thanks!!
